[jinbesan] iP

jinbesan opened the pull request:
# Duke Task List 😳 
## A simple text-based app to manage tasks
> "[GitHub Flavored Markdown](https://guides.github.com/features/mastering-markdown/) (and Markdown in general) is useful in many places when using GitHub e.g., issue tracker, PR reviews, writing documentation. The aim of this task is to ensure that you are sufficiently familiar with the GFMD syntax."
- Separate tasks between:
  - Todos
  - Events
  - Deadlines

1. **Download** the .jar file
2. **Run** it (double click), alternatively **run** `java -jar Duke.jar` in the same folder as the jar file 
3. **Follow** the text based instructions to add tasks to the list!

### This pull request includes:
- [x] a heading
- [x] a bullet list
- [x] a numbered list
- [x] a fenced code block (with syntax highlighting)
- [x] a task list ~~Like this one!~~
- [x] an emoji
- [x] a blockquote
- [x] a [hyper link](https://en.wikipedia.org/wiki/Hyperlink)
- [x] inline code
- [x] some text formatting: bold, italic, stikethrough etc.

``` java
public static void main(String[] args) {
    System.out.println("hello duke");
}
```

Commit:
Add Gradle support

jinbesan committed:
Added Greet, Echo, Exit

jinbesan committed:
Fixed bugs under Level-1

jinbesan committed:
Added add item to list, and list

jinbesan committed:
Added Task class as abstraction for Task (Task.java)
Added mark and unmark functions for list

jinbesan committed:
Added subclasses of Task: Todo, Deadline, Event.
User can now list whether the Task is Todo, Deadline, or Event and specify a time for the latter two.

jinbesan committed:
Duke.java:
Reformatted code to make it neater and streamlined.

Added DukeException.java:
provide Duke-specific exception handling

jinbesan committed:
Duke.java:
Added "remove" command
Added additional exceptions where time was not specified for Deadline or Event

jinbesan committed:
Added text-ui testing

jinbesan committed:
Restructured Duke.java into more classes to implement OOP classes
gitignore data folder
added storage function (save and load) (Storage.java)

jinbesan committed:
Organized classes into packages

jinbesan committed:
Changed Deadline to only accept valid dates

jinbesan committed:
Merge remote-tracking branch 'origin/add-gradle-support'

jinbesan committed:
Added ParserTest.java to test Parser.parse
UiTest.java to test ui.showDeleteTask

jinbesan committed:
Edited build.gradle to create the correct jar file

jinbesan committed:
Added Javadocs

jinbesan committed:
Revert "Added Javadocs"

This reverts commit 21d669d0012bc26e16aaf83311cc1036483dda23.

jinbesan committed:
Revert "Revert "Added Javadocs""

This reverts commit 162d05fccdf71c845ba5faebdce4e796df4f8478.

jinbesan committed:
Changed code to adhere to Coding Standards

jinbesan committed:
Added a Find function to find tasks which description contains a specified keyword.

jinbesan committed:
Merge branch 'branch-Level-9' into branch-A-JavaDoc

jinbesan committed:
Merge branch 'branch-A-JavaDoc' into branch-A-CodingStandard

Mr-Teal commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,43 @@
+        while (!isExit) {
+            try {
+                String fullCommand = ui.readCommand();
+                ui.showSepLine(); // show the divider line ("_______")
```
Clear comment here.

Mr-Teal commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,103 @@
+
+    }
+
+    public static Command parse(String cmd) throws DukeException{
```
I like the clear separation of different command types in this method.

junlee1991 replied:
I like the clear separation of different command types in this method.

Mr-Teal commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,84 @@
+        this.showHelp();
+    }
+
+    public void showHelp() {
```
Good showHelp() method here.

junlee1991 replied:
Good showHelp() method here.

Mr-Teal reviewed (commented):
The program seems quite good. It also has some nice extra parts I like. Nice!

jinbesan committed:
Added CheckStyle plugin to IntelliJ

junlee1991 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,43 @@
+package duke;
```
Overall, I found your code easy to read and understand! Well done!

junlee1991 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,70 @@
+        return tasks.size();
+    }
+
+    public String createTaskSaveData() {
```
I like how the code is abstracted in methods!

junlee1991 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,26 @@
+    public abstract String save();
+
+    @Override
+    public String toString(){
```
I think the Egyptian bracket is supposed to have 1 spacing.

jinbesan committed:
Added javafx UI to application, main application in Duke and launched from Launcher. void print commands in Ui are now replaced to return strings.

jinbesan committed:
Merge branch 'branch-A-CheckStyle' into branch-Level-10

# Conflicts:
#	build.gradle

jinbesan committed:
Merge branch 'branch-Level-10'

homuntan02 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,115 @@
+        String command = tokens[0];
+
+        // List Tasks
+        if (command.equals("list")) {
```
Maybe instead of if statements, you can use switch cases.

homuntan02 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,151 @@
+                "Exit Duke: exit";
+    }
+
+    /**
```
Great additional method.

homuntan02 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,84 @@
+            return new Todo(taskData[1].equals("1"), taskData[2]);
+        } else if (taskData[0].equals("E")) {
+            return new Event(taskData[1].equals("1"), taskData[2], taskData[3]);
+        } else if (taskData[0].equals("D")) {
```
Maybe implement switch cases here as well.

homuntan02 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,115 @@
+        // Remove task
+        if (command.equals("remove")) {
+            if (tokens.length != 2) {
+                throw new DukeException("Please enter an index to remove!");
+            }
+            try {
+                int index = Integer.parseInt(tokens[1]);
+                return new DeleteCommand(index);
+            }
+            catch (NumberFormatException nfe) {
+                throw new DukeException("Index must be an integer!");
```
I like how you seperated different exceptions here.

homuntan02 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,36 @@
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage){
+        tasks.addTask(task);
+        storage.save(tasks);
```
Good use of Storage, saving as task are added.

junlee1991 reviewed (commented):
looks good

jinbesan committed:
no message

jinbesan committed:
Add Java asserts

No use of assert statements in program code

Insufficient testing for bugs caused by the programmer in the code.

Let's add Java asserts to define assumptions in the program code, like
making sure the Command is not null.

Using Java asserts allows the runtime to verify these assumptions and
help find bugs in the code.

jinbesan committed:
Improve code quality

jinbesan committed:
Merge pull request #1 from jinbesan/branch-A-Assertions

Add Java asserts

jinbesan committed:
Merge pull request #2 from jinbesan/branch-A-CodeQuality

Improve code quality

jinbesan committed:
Add specific help command for different commands

Help message only displayed when app starts, and all of the information
is in that 1 message.

There is a need to scroll all the way to the beginning if the user
forgets how to use a command. It is also very cluttered with
information on how to use the commands.

Let's add specific help commands like help [command], and change
help prompt on app start to something more concise.

Keeping the starting help prompt tidy will allow users to quickly
navigate to what they want, and help requests for specific commands
will allow more detailed explanations and examples on how to use
the command.

jinbesan committed:
Changed UI to become wider and taller for better visibility
Added App Icon
Changed App personality

jinbesan committed:
Updated user guide, added product screenshot

jinbesan committed:
no message
